shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -2594,169 +2611,211 @@ impl KafkaSinkCluster {
+                            // and fall back to use all UP shotover nodes out of the rack otherwise.
+                            let (shotover_nodes_in_rack, shotover_nodes_out_of_rack) =
+                                partition_shotover_nodes_by_rack(
+                                    &up_shotover_nodes,
+                                    &rack.as_ref(),
+                                );
+                            let unique_broker_ids = collect_unique_broker_ids(
+                                shotover_nodes_in_rack,
+                                shotover_nodes_out_of_rack,
+                            );
+                            shotover_isr_nodes.extend(unique_broker_ids);
+                        } else {
+                            // If the cluster detects the broker is down it will not be included in the metadata list of brokers.
+                            // In that case we wont find a broker and we should just skip this broker.
+                        }
+                    }
+                    partition.isr_nodes = shotover_isr_nodes;
 
-                // TODO: handle this properly, for now its better to just clear it than do nothing
-                partition.offline_replicas.clear();
+                    // TODO: handle this properly, for now its better to just clear it than do nothing
+                    partition.offline_replicas.clear();
+                }
             }
-        }
 
-        if let Some(controller_node) = self
-            .nodes
-            .iter()
-            .find(|node| node.broker_id == metadata.controller_id)
-        {
-            // If broker has no rack - use the first shotover node
-            // If broker has rack - use the first shotover node with the same rack
-            // This is deterministic because the list of shotover nodes is sorted.
-            if let Some(shotover_node) = self.shotover_nodes.iter().find(|shotover_node| {
-                controller_node
-                    .rack
-                    .as_ref()
-                    .map(|rack| rack == &shotover_node.rack)
-                    .unwrap_or(true)
-            }) {
-                metadata.controller_id = shotover_node.broker_id;
+            if let Some(controller_node_rack) = self
+                .nodes
+                .iter()
+                .find(|node| node.broker_id == metadata.controller_id)
+                .map(|x| x.rack.clone())
+            {
+                // If broker has no rack - use the first UP shotover node.
+                // If broker has rack - use the first UP shotover node with the same rack if available,
+                // and fall back to use the first UP shotover node out of the rack otherwise.
+                // This is deterministic because the list of UP shotover nodes is sorted.
+                let (shotover_nodes_in_rack, shotover_nodes_out_of_rack) =
+                    partition_shotover_nodes_by_rack(
```

Comment:
confirm if partition ensures the original order

Reply:
If my understanding is right, partition should retain the original order as it adds to left and right by `extend_one()` by looking at the [source code](https://doc.rust-lang.org/1.81.0/src/core/iter/traits/iterator.rs.html#2186-2190).